File: kimi-k3/reference/python/kimi_k3_ref/expert_ffn.py
"""Expert FFN + LatentMoE path matching KimiBlockSparseMLP / KimiSparseMoeBlock."""

from __future__ import annotations

import numpy as np

from .moe_gate import moe_gate
from .rmsnorm import rmsnorm
from .situ import situ_and_mul


def linear(x: np.ndarray, weight: np.ndarray) -> np.ndarray:
    """y = x @ W.T  with W shaped (out, in), matching nn.Linear bias=False."""
    return (np.asarray(x, dtype=np.float32) @ np.asarray(weight, dtype=np.float32).T).astype(
        np.float32
    )


def expert_ffn(
    hidden: np.ndarray,
    w1: np.ndarray,
    w2: np.ndarray,
    w3: np.ndarray,
    beta: float = 4.0,
    linear_beta: float | None = 25.0,
) -> np.ndarray:
    """
    Official KimiBlockSparseMLP with situ:
        gate_up = cat(w1(x), w3(x), dim=-1)
        y = situ(gate_up)
        return w2(y)
    """
    gate = linear(hidden, w1)
    up = linear(hidden, w3)
    gate_up = np.concatenate([gate, up], axis=-1)
    y = situ_and_mul(gate_up, beta=beta, linear_beta=linear_beta)
    return linear(y, w2)


def dense_mlp(
    hidden: np.ndarray,
    gate_proj: np.ndarray,
    up_proj: np.ndarray,
    down_proj: np.ndarray,
    beta: float = 4.0,
    linear_beta: float | None = 25.0,
) -> np.ndarray:
    """Shared / dense MLP matching KimiMLP with situ."""
    gate = linear(hidden, gate_proj)
    up = linear(hidden, up_proj)
    gate_up = np.concatenate([gate, up], axis=-1)
    y = situ_and_mul(gate_up, beta=beta, linear_beta=linear_beta)
    return linear(y, down_proj)


def moe_infer(
    x: np.ndarray,
    topk_ids: np.ndarray,
    topk_weight: np.ndarray,
    expert_weights: list[tuple[np.ndarray, np.ndarray, np.ndarray]],
    beta: float = 4.0,
    linear_beta: float | None = 25.0,
) -> np.ndarray:
    """
    Official moe_infer (simplified, EP size=1):
    - scatter tokens to experts by topk_ids
    - run each expert
    - weighted sum by topk_weight
    """
    x = np.asarray(x, dtype=np.float32)
    n, h = x.shape
    top_k = topk_ids.shape[1]
    # Accumulate weighted expert outputs per token
    out = np.zeros((n, h), dtype=np.float32)
    for t in range(n):
        acc = np.zeros((h,), dtype=np.float32)
        for k in range(top_k):
            e = int(topk_ids[t, k])
            w1, w2, w3 = expert_weights[e]
            y = expert_ffn(x[t : t + 1], w1, w2, w3, beta=beta, linear_beta=linear_beta)[0]
            acc += topk_weight[t, k] * y
        out[t] = acc
    return out


def latent_moe_forward(
    hidden_states: np.ndarray,
    router_weight: np.ndarray,
    router_bias: np.ndarray,
    top_k: int,
    expert_weights: list[tuple[np.ndarray, np.ndarray, np.ndarray]],
    routed_down: np.ndarray,
    routed_up: np.ndarray,
    routed_norm_weight: np.ndarray | None,
    shared_gate: np.ndarray,
    shared_up: np.ndarray,
    shared_down: np.ndarray,
    rms_eps: float = 1e-5,
    beta: float = 4.0,
    linear_beta: float | None = 25.0,
    routed_scaling_factor: float = 1.0,
) -> np.ndarray:
    """
    LatentMoE forward matching KimiSparseMoeBlock inference:
      identity = h
      route on h
      h' = down_proj(h)
      y = moe_infer(h', ...)
      y = rmsnorm(y); y = up_proj(y)
      y = y + shared_experts(identity)
    """
    identity = np.asarray(hidden_states, dtype=np.float32)
    orig_shape = identity.shape
    topk_idx, topk_weight = moe_gate(
        identity,
        router_weight,
        router_bias,
        top_k=top_k,
        routed_scaling_factor=routed_scaling_factor,
        renormalize=True,
    )
    flat = identity.reshape(-1, identity.shape[-1])
    latent = linear(flat, routed_down)
    y = moe_infer(latent, topk_idx, topk_weight, expert_weights, beta=beta, linear_beta=linear_beta)
    if routed_norm_weight is not None:
        y = rmsnorm(y, routed_norm_weight, eps=rms_eps)
    y = linear(y, routed_up)
    shared = dense_mlp(flat, shared_gate, shared_up, shared_down, beta=beta, linear_beta=linear_beta)
    y = y + shared
    return y.reshape(orig_shape).astype(np.float32)

File: kimi-k3/rust/src/expert_ffn.rs
use crate::moe_gate::moe_gate;
use crate::rmsnorm::rmsnorm;
use crate::situ::situ_and_mul;

/// `y = x @ W.T` with W shaped (out, in).
pub fn linear(x: &[f32], rows: usize, in_dim: usize, weight: &[f32], out_dim: usize, out: &mut [f32]) {
    assert_eq!(x.len(), rows * in_dim);
    assert_eq!(weight.len(), out_dim * in_dim);
    assert_eq!(out.len(), rows * out_dim);
    for r in 0..rows {
        for o in 0..out_dim {
            let mut s = 0.0f32;
            let w = &weight[o * in_dim..(o + 1) * in_dim];
            let xr = &x[r * in_dim..(r + 1) * in_dim];
            for j in 0..in_dim {
                s += xr[j] * w[j];
            }
            out[r * out_dim + o] = s;
        }
    }
}

pub fn expert_ffn(
    x: &[f32],
    rows: usize,
    hidden: usize,
    w1: &[f32],
    w2: &[f32],
    w3: &[f32],
    inter: usize,
    beta: f32,
    linear_beta: Option<f32>,
    out: &mut [f32],
) {
    let mut gate = vec![0.0f32; rows * inter];
    let mut up = vec![0.0f32; rows * inter];
    linear(x, rows, hidden, w1, inter, &mut gate);
    linear(x, rows, hidden, w3, inter, &mut up);
    let mut gate_up = vec![0.0f32; rows * (2 * inter)];
    for r in 0..rows {
        for i in 0..inter {
            gate_up[r * 2 * inter + i] = gate[r * inter + i];
            gate_up[r * 2 * inter + inter + i] = up[r * inter + i];
        }
    }
    let mut act = vec![0.0f32; rows * inter];
    situ_and_mul(&gate_up, 2 * inter, beta, linear_beta, &mut act);
    linear(&act, rows, inter, w2, hidden, out);
}

pub fn dense_mlp(
    x: &[f32],
    rows: usize,
    hidden: usize,
    gate_proj: &[f32],
    up_proj: &[f32],
    down_proj: &[f32],
    inter: usize,
    beta: f32,
    linear_beta: Option<f32>,
    out: &mut [f32],
) {
    expert_ffn(
        x, rows, hidden, gate_proj, down_proj, up_proj, inter, beta, linear_beta, out,
    );
}

pub fn latent_moe_forward(
    hidden_states: &[f32],
    batch: usize,
    seq: usize,
    hidden: usize,
    router_weight: &[f32],
    router_bias: &[f32],
    num_experts: usize,
    top_k: usize,
    experts_w1: &[Vec<f32>],
    experts_w2: &[Vec<f32>],
    experts_w3: &[Vec<f32>],
    latent: usize,
    inter: usize,
    routed_down: &[f32],
    routed_up: &[f32],
    routed_norm: Option<&[f32]>,
    shared_gate: &[f32],
    shared_up: &[f32],
    shared_down: &[f32],
    rms_eps: f32,
    beta: f32,
    linear_beta: Option<f32>,
    out: &mut [f32],
) {
    let n = batch * seq;
    assert_eq!(hidden_states.len(), n * hidden);
    assert_eq!(out.len(), n * hidden);

    let mut topk_idx = vec![0i64; n * top_k];
    let mut topk_weight = vec![0.0f32; n * top_k];
    moe_gate(
        hidden_states,
        n,
        hidden,
        router_weight,
        num_experts,
        router_bias,
        top_k,
        1.0,
        true,
        &mut topk_idx,
        &mut topk_weight,
    );

    let mut latent_x = vec![0.0f32; n * latent];
    linear(hidden_states, n, hidden, routed_down, latent, &mut latent_x);

    let mut y = vec![0.0f32; n * latent];
    for t in 0..n {
        let xt = &latent_x[t * latent..(t + 1) * latent];
        let mut acc = vec![0.0f32; latent];
        for k in 0..top_k {
            let e = topk_idx[t * top_k + k] as usize;
            let mut tmp = vec![0.0f32; latent];
            expert_ffn(
                xt,
                1,
                latent,
                &experts_w1[e],
                &experts_w2[e],
                &experts_w3[e],
                inter,
                beta,
                linear_beta,
                &mut tmp,
            );
            let w = topk_weight[t * top_k + k];
            for j in 0..latent {
                acc[j] += w * tmp[j];
            }
        }
        y[t * latent..(t + 1) * latent].copy_from_slice(&acc);
    }

    if let Some(nw) = routed_norm {
        let mut yn = vec![0.0f32; n * latent];
        rmsnorm(&y, nw, rms_eps, &mut yn);
        y = yn;
    }

    let mut routed = vec![0.0f32; n * hidden];
    linear(&y, n, latent, routed_up, hidden, &mut routed);

    let mut shared = vec![0.0f32; n * hidden];
    dense_mlp(
        hidden_states,
        n,
        hidden,
        shared_gate,
        shared_up,
        shared_down,
        inter,
        beta,
        linear_beta,
        &mut shared,
    );

    for i in 0..n * hidden {
        out[i] = routed[i] + shared[i];
    }
}
